Missing link with -latomic on ARM: error: undefined reference to '__atomic_fetch_add_8'

### How to reproduce ###
<!-- Steps to reproduce the behavior: -->

Run `cmake` then `make` on an ARM system.

### Actual behavior ###
<!-- What happened? -->

```
[ 69%] Linking CXX executable ccache
src/libccache_lib.a(InodeCache.cpp.o):InodeCache.cpp:function InodeCache::acquire_bucket(unsigned int): error: undefined reference to '__atomic_fetch_add_8'
src/libccache_lib.a(InodeCache.cpp.o):InodeCache.cpp:function InodeCache::acquire_bucket(Digest const&): error: undefined reference to '__atomic_fetch_add_8'
src/libccache_lib.a(InodeCache.cpp.o):InodeCache.cpp:function InodeCache::get(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InodeCache::ContentType, Digest&, int*): error: undefined reference to '__atomic_fetch_add_8'
src/libccache_lib.a(InodeCache.cpp.o):InodeCache.cpp:function InodeCache::get(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InodeCache::ContentType, Digest&, int*): error: undefined reference to '__atomic_load_8'
src/libccache_lib.a(InodeCache.cpp.o):InodeCache.cpp:function InodeCache::get(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InodeCache::ContentType, Digest&, int*): error: undefined reference to '__atomic_load_8'
src/libccache_lib.a(InodeCache.cpp.o):InodeCache.cpp:function InodeCache::get(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InodeCache::ContentType, Digest&, int*): error: undefined reference to '__atomic_load_8'
src/libccache_lib.a(InodeCache.cpp.o):InodeCache.cpp:function InodeCache::get(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InodeCache::ContentType, Digest&, int*): error: undefined reference to '__atomic_fetch_add_8'
src/libccache_lib.a(InodeCache.cpp.o):InodeCache.cpp:function InodeCache::get_hits(): error: undefined reference to '__atomic_load_8'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/ccache.dir/build.make:108: ccache] Error 1
make[1]: *** [CMakeFiles/Makefile2:252: CMakeFiles/ccache.dir/all] Error 2
make: *** [Makefile:183: all] Error 2
```

### Expected behavior ###
<!-- What did you expect to happen? -->

The link should include `-latomic` to find the missing symbol definitions.

### Proposed workaround ###
Adding `atomic` to the `target_link_libraries(ccache …)` line in `CMakeLists.txt` and to the `target_link_libraries(unittest …)` line in `unittest/CMakeLists.txt` allows the build to complete, but perhaps a proper fix should do this conditionally, as some platforms may not require it.

### Environment ###
<!-- Which ccache version did you use? -->

ccache commit 015bfd80b526dd00708ae6930e65229a8f3c4fea

<!-- If applicable, add more info such as compiler, OS and file system below: -->

```
# uname -a
Linux raspberrypi 5.4.71+ #1352 Fri Oct 16 17:59:06 BST 2020 armv6l ARMv6-compatible processor rev 7 (v6l) BCM2835 GNU/Linux

# gcc --version | head -n1
gcc (Gentoo 9.3.0-r1 p3) 9.3.0
```

--- a/cmake/StandardSettings.cmake
+++ b/cmake/StandardSettings.cmake
@@ -47,6 +47,8 @@ if(CMAKE_CXX_COMPILER_ID MATCHES "^GNU|(Apple)?Clang$")
       INTERFACE -fsanitize=${SANITIZER})
   endforeach()
 
+  include(StdAtomic)
+
 elseif(MSVC)
   target_compile_options(standard_settings INTERFACE /std:c++latest /Zc:preprocessor /Zc:__cplusplus /D_CRT_SECURE_NO_WARNINGS)
 endif()

--- a/cmake/StdAtomic.cmake
+++ b/cmake/StdAtomic.cmake
@@ -1,0 +1,31 @@
+# Check if std::atomic needs -latomic
+
+include(CheckCXXSourceCompiles)
+
+set(CMAKE_REQUIRED_FLAGS ${CMAKE_CXX11_STANDARD_COMPILE_OPTION})
+set(
+  check_std_atomic_source_code
+  [=[
+    #include <atomic>
+    int main()
+    {
+      std::atomic<long long> x;
+      (void)x.load();
+      return 0;
+    }
+  ]=])
+
+check_cxx_source_compiles("${check_std_atomic_source_code}" std_atomic_without_libatomic)
+
+if(NOT std_atomic_without_libatomic)
+  set(CMAKE_REQUIRED_LIBRARIES atomic)
+  check_cxx_source_compiles("${check_std_atomic_source_code}" std_atomic_with_libatomic)
+  set(CMAKE_REQUIRED_LIBRARIES)
+  if(NOT std_atomic_with_libatomic)
+    message(FATAL_ERROR "Toolchain doesn't support std::atomic with nor without -latomic")
+  else()
+    target_link_libraries(standard_settings INTERFACE atomic)
+  endif()
+endif()
+
+set(CMAKE_REQUIRED_FLAGS)

